pkg/output/table.go

```diff
@@ -0,0 +1,322 @@
+		}
+		if column == nil {
+			column = b.defaultColumn(columnName)
+		}
+		table.columns[i] = column
+	}
+
+	// Return the table:
+	result = table
+	return
+}
+
+// loadColumnYAML copies the column data from the YAML document to the object.
+func (b *TableBuilder) loadColumn(i int, columnData *columnYAML) (result *Column, err error) {
+	// Check that the name of the column has been specified:
+	if columnData.Name == nil || *columnData.Name == "" {
+		err = fmt.Errorf(
+			"column %d of table '%s' doesn't have a name",
+			i, b.name,
+		)
+		return
+	}
+	columnName := *columnData.Name
+
+	// Create an initially empty column:
+	column := b.defaultColumn(columnName)
+
+	// Copy the data from the YAML document:
+	if columnData.Header != nil {
+		column.header = *columnData.Header
+	}
+	if columnData.Width != nil {
+		column.width = *columnData.Width
+	}
+
+	// Return the column:
+	result = column
+	return
+}
+
+// defaultColumn creates a default column description for the given column name.
+func (b *TableBuilder) defaultColumn(columnName string) *Column {
+	return &Column{
+		name:   columnName,
+		header: b.defaultColumnHeader(columnName),
+		width:  b.defaultColumnWidth(columnName),
+	}
+}
+
+// defaultColumnHeader returns the default value for the header of the given column name.
+func (b *TableBuilder) defaultColumnHeader(columnName string) string {
+	columnHeader := columnName
+	columnHeader = strings.ReplaceAll(columnHeader, ".", " ")
+	columnHeader = strings.ReplaceAll(columnHeader, "_", " ")
+	columnHeader = strings.ToUpper(columnHeader)
+	return columnHeader
+}
+
+// defaultColumnWidth returns the default width for the given column name.
+func (b *TableBuilder) defaultColumnWidth(columnName string) int {
+	return len(columnName)
+}
+
+// WriteColumns writes a row of a table using the given values.
+func (t *Table) WriteColumns(columnValues []interface{}) error {
+	// Check that the number of values matches the number of columns:
+	valueCount := len(columnValues)
+	columnCount := len(t.columns)
+	if valueCount != columnCount {
+		return fmt.Errorf(
+			"table '%s' has %d columns, but %d values have been given",
+			t.name, columnCount, valueCount,
+		)
+	}
+
+	// Prepare a buffer to write the columns (sum of the widths of the columns plus two
+	// characters to separate columns, and the new line):
+	tableWidth := 2 * columnCount
+	for _, column := range t.columns {
+		tableWidth += column.width
+	}
+	var rowBuffer bytes.Buffer
+	rowBuffer.Grow(tableWidth)
+
+	// Write the values while trimming or padding to adjust to the desired sizes:
+	for i, columnValue := range columnValues {
+		if i > 0 {
+			rowBuffer.WriteString("  ")
+		}
+		var columnText string
+		if columnValue != nil {
+			columnText = fmt.Sprintf("%v", columnValue)
+		} else {
+			columnText = "NONE"
+		}
+		actualWidth := len(columnText)
+		desiredWidth := t.columns[i].width
+		switch {
+		case actualWidth > desiredWidth:
+			rowBuffer.WriteString(columnText[0:desiredWidth])
```

Comment:
This still silently truncates texts that don't fit.  
IMHO that does not solve #226 / https://issues.redhat.com/browse/SDA-4246, the fundanmental question there is not "good defaults" but "what to do when it didn't fit".

However, it's an improvement on current one-width-fits-all behavior, and improving the truncation behavior is an orthogonal change that will be easier once all commands use `printer`, so :+1:

Reply:
One of the things that we can do once we have metadata for tables/columns is introduce rules to decide what to do when they don't fit. For example, we could have something like this:

```yaml
columns:
- name: api.url
  trim: false
```

That would tell the CLI that the "api.url" column should never be trimmed. It could then use this information to move the column to the end of the table, and never trim it. Obviously this can't be done if there are multiple columns with `trim: false`, but the point is that it makes it possible to make the display logic much more intelligent.

Another possibility would be to at least mark those columns to explicitly indicate that they have been trimmed:

```yaml
columns:
- name: api.url
  ellipsis: true
```

That hypothetical configuration would mean that the value should be terminaled with `…` to visually indicate that it has been trimmed:

```
$ ocm list clusters --columns id,api.url,name
ID                                API URL                       NAME
1i6anlpc6278j4oa5nbp8ik69dl6h53j  https://api.osde2e-hcltn.3a…  osde2e-hcltn
```

So, I agree that this doesn't solve all the problems, it is just an starting point.